# Аргументы функций

Немного подробнее про аргументы функций

## Оглавление
- Что такое аргументы функций и зачем они нужны
- Как создать, как использовать
- Стандартные значения аргументов
- Именованные аргументы
- Как сочетаются разные типы аргументов

## Что такое аргументы функций и зачем они нужны

Вы уже наверняка знакомы с функциями - это блок кода, который выполняется при вызове его именованной области(функции).

Аргументы функций призваны расширять возможность функций работать с данными, которые появляются до моменты их выполнения. Например, при написании программы-калькулятора, который воспроизводит базовые арифметические операции над двумя числами, эти самые два числа создаются до момента вызова функции, а уже затем появляется возможность выбора необходимого действия - сложения, вычитания и так далее.

Так вот, эти числа должны быть переданы в функцию, чтобы она могла произвести работу над ними. Синтаксис выглядит так (пример с функцией сложения двух чисел):

```python
def add(num1, num2):
    print(num1 + num2)

a = int(input())
b = int(input())
add(a, b)
```

В данном случае, мы создаем переменные a и b с числовыми значениями, затем передаем их в функцию.

Сама функция принимает эти два значения и записывает их по очереди в переменные num1 и num2 - то есть те самые ***аргументы функции***.

Без них функция не могла бы принять значения извне.

## Как создать, как использовать

Создание аргументов происходит вместе с созданием функции. Аргументы прописываются в скобках после названия функции:

```python
def function(argument1, argument2, argument3, ...):
```

Аргументов функций может как не быть, так и быть столько, сколько нужно (два, пять, десять, сто, миллион).

Используются аргументы функций при вызове самой функции передачей в нее значений в том порядке, в котором были прописаны функции изначально - то есть, иными словами, порядок важен!

Важно также учитывать, что если аргументов, например, 3, то и передаваемых значений должно быть 3 - ни больше, ни меньше.

Тем не менее, это правило "ломается" с аргументами, у которых есть стандартные значения об этом в следующем пункте.

## Стандартные значения аргументов

Возможно также создание таких аргументов, у которых есть стандартное значение(значение по умолчанию):

```python
def function(
    argument1 = "hello", 
    argument2 = 123, 
    argument3 = False):
```

В данном примере у аргументов есть некоторое значение, которое будет присвоено им, если мы не передадим свои значения в функцию. Так, например, если произвести "пустой" вызов функции:

```python
function()
```

То каждый аргумент "присвоит" себе то значение, которое мы прописали в созданной сверху функции, то есть в **argument1** попадет строка **"hello"**, в **argument2** число **123**, а в **argument3** булево значение **False**.

Если же мы передадим несколько значений:

```python
function(1, 'abc')
```

То они сохранятся в аргументы, на которые "их хватит" по количеству. То есть, если у функции три аргумента со стандартными значениями, а мы передаем два значения в функцию, то первые два аргумента примут наши значения, а третий "присвоит" себе свое стандартное значение **False**. Иными словами, порядок все так же важен, но количество передаваемых значений уже может быть меньше.

Если в такую функцию передать больше значений, чем количество аргументов, то это все так же вызовет ошибку.

Чтобы "сломать" правило с порядком аргументов, то нужно прибегнуть к работе с именованными аргументами.

## Именованные аргументы

По сути, это все те же аргументы о которых речь шла ранее, но порядок передачи значений в функцию может изменяться вследствие передачи значения в конкретный аргумент при вызове:

```python
def function(a, b, c):
    print(a, b, c)

function(b=3, c=2, a=1)
```

При запуске в выводе получится следующее:
```
>>> 1 3 2
```

То есть, несмотря на порядок передачи значения, передаем мы их в конкретный аргумент, куда они и сохраняются.

Указать сторонний аргумент не выйдет, такая практика вызовет ошибку:

```python
def function(a, b, c):
    print(a, b, c)

function(D=3, c=2, a=1)
```

```
TypeError: function() got an unexpected keyword argument 'D'
```

## Как сочетаются разные типы аргументов

При работе с различными аргументами и передачами значений важно знать следующие правила:

#### 1) В функции сначала обязаны идти аргументы без стандартных значений, а уже затем с ними - не наоборот и не в случайном порядке.

Неправильно:
```python
def function(a=10, b, c=3):
    ...
```
```python
def function(a=10, b=2, c):
    ...
```
До обычного аргумента стоит тот, у которого есть стандартное значение - это некорректно, вызовет ошибку.
```
SyntaxError: non-default argument follows default argument
```

<br>

Правильно:
```python
def function(a, b, c=3):
    ...
```
```python
def function(a, b=2, c=1):
    ...
```
Сначала идут обычные аргументы, а уже затем со стандартными значениями - корректная работа с аргументами разных типов.

#### 2) В вызов функции должны сначала передаваться значения без именованных аргументов, и только затем с ними.

Как и в первом правиле, "перемешивать" передачу значений нельзя - сначала всегда идут значения для обычных аргументов.

Неправильно:
```python
function(a=521, 'abc')
```
```python
function(a=1, b='abc', 888)
```
До обычной передачи значений идут передачи через именованные аргументы - некорректное использование функционала языка вызовет ошибку.
```
SyntaxError: positional argument follows keyword argument *
```
*- как видно из ошибки, "обычные" аргументы называются позиционными. Чаще всего их так и называют в документациях языка и прочих библиотек.

<br>

Правильно:
```python
function(a=521, b='abc')
```
```python
function(1, 'abc', c=888)
```
Значения обычных и именованных аргументов не перемешаны, идут в строгом порядке, следуя правилу - корректная работа с функционалом языка.
